File: src/2025-02-27-deepclaude37-rust/javascript/exercises/practice/scale-generator/scale-generator.js
//
// This is only a SKELETON file for the 'Scale Generator' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class Scale {
  constructor(tonic) {
    this.tonic = this.normalizeTonic(tonic);
    this.keySignature = this.determineKeySignature(this.tonic);
    this.sharpSequence = ['A', 'A#', 'B', 'C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#'];
    this.flatSequence = ['A', 'Bb', 'B', 'C', 'Db', 'D', 'Eb', 'E', 'F', 'Gb', 'G', 'Ab'];
    this.sequence = this.keySignature === 'sharps' ? this.sharpSequence : this.flatSequence;
    this.startIndex = this.sequence.indexOf(this.tonic);
    if (this.startIndex === -1) {
      throw new Error('Invalid tonic');
    }
  }

  normalizeTonic(tonic) {
    const notePart = tonic[0]; // Preserve case to determine major/minor
    const accidental = tonic.slice(1).toLowerCase();
    return notePart + accidental;
  }

  determineKeySignature(normalizedTonic) {
    // Natural keys
    if (normalizedTonic === 'C' || normalizedTonic === 'a') {
      return 'sharps';
    }
    
    // Check if major (uppercase) or minor (lowercase)
    const isMajor = normalizedTonic[0] === normalizedTonic[0].toUpperCase();
    
    if (isMajor) {
      // Major keys
      const majorSharps = new Set(['G', 'D', 'A', 'E', 'B', 'F#']);
      if (majorSharps.has(normalizedTonic)) {
        return 'sharps';
      }
      
      const majorFlats = new Set(['F', 'Bb', 'Eb', 'Ab', 'Db', 'Gb']);
      if (majorFlats.has(normalizedTonic)) {
        return 'flats';
      }
    } else {
      // Minor keys
      const minorSharps = new Set(['e', 'b', 'f#', 'c#', 'g#', 'd#']);
      if (minorSharps.has(normalizedTonic)) {
        return 'sharps';
      }
      
      const minorFlats = new Set(['d', 'g', 'c', 'f', 'bb', 'eb']);
      if (minorFlats.has(normalizedTonic)) {
        return 'flats';
      }
    }
    
    throw new Error('Invalid tonic');
  }

  chromatic() {
    const scale = [];
    for (let i = 0; i < 12; i++) {
      const index = (this.startIndex + i) % 12;
      // Ensure all notes are uppercase (except 'b' in flats)
      const note = this.sequence[index];
      scale.push(note[0].toUpperCase() + note.slice(1));
    }
    return scale;
  }

  interval(intervals) {
    let currentIndex = this.startIndex;
    const scale = [this.sequence[currentIndex]];
    for (const interval of intervals) {
      let steps;
      switch (interval) {
        case 'm':
          steps = 1;
          break;
        case 'M':
          steps = 2;
          break;
        case 'A':
          steps = 3;
          break;
        default:
          throw new Error(`Invalid interval: ${interval}`);
      }
      currentIndex = (currentIndex + steps) % 12;
      scale.push(this.sequence[currentIndex]);
    }
    return scale;
  }
}

File: src/2025-02-27-deepclaude37-rust/python/exercises/practice/scale-generator/scale_generator.py
class Scale:
    def __init__(self, tonic):
        # Determine if the tonic is for a major or minor scale based on case
        is_minor = tonic[0].islower()
        
        # Normalize the tonic for scale lookup
        note_part = tonic[0].upper() + tonic[1:]
        
        # Define key signature sets based on the table in the instructions
        major_sharps = {'C', 'G', 'D', 'A', 'E', 'B', 'F#'}
        major_flats = {'F', 'Bb', 'Eb', 'Ab', 'Db', 'Gb'}
        
        minor_sharps = {'A', 'E', 'B', 'F#', 'C#', 'G#', 'D#'}
        minor_flats = {'D', 'G', 'C', 'F', 'Bb', 'Eb'}
        
        # Determine whether to use sharps or flats based on the tonic and its mode
        if is_minor:
            if note_part in minor_sharps:
                self.use_sharps = True
            else:
                self.use_sharps = False
        else:  # major
            if note_part in major_sharps:
                self.use_sharps = True
            else:
                self.use_sharps = False
        
        # Store the normalized tonic for scale generation
        self.tonic = note_part

    def chromatic(self):
        # Define the two versions of the chromatic scale
        sharps_order = ['A', 'A#', 'B', 'C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#']
        flats_order = ['A', 'Bb', 'B', 'C', 'Db', 'D', 'Eb', 'E', 'F', 'Gb', 'G', 'Ab']
        
        # Select appropriate scale based on tonic
        selected_order = sharps_order if self.use_sharps else flats_order
        
        # Rotate the scale to start with the tonic
        index = selected_order.index(self.tonic)
        return selected_order[index:] + selected_order[:index]

    def interval(self, intervals):
        # Get the chromatic scale starting with the tonic
        chromatic_scale = self.chromatic()
        
        # Start with the tonic
        result = [chromatic_scale[0]]
        current_index = 0
        
        # Map intervals to steps in chromatic scale
        interval_steps = {'m': 1, 'M': 2, 'A': 3}
        
        # Build the scale according to interval pattern
        for interval in intervals:
            current_index = (current_index + interval_steps[interval]) % 12
            result.append(chromatic_scale[current_index])
            
        return result
